sae/blocks.go

```diff
@@ -104,74 +101,148 @@ func (vm *VM) buildBlock(
+				zap.Error(err),
+			)
+			return nil, fmt.Errorf("starting worst-case state for block %d: %v", b.Height(), err)
+		}
+		for i, tx := range b.Transactions() {
+			if err := state.ApplyTx(tx); err != nil {
+				log.Warn("Could not apply tx during historical worst case calculation",
+					zap.Int("tx_index", i),
+					zap.Stringer("tx_hash", tx.Hash()),
+					zap.Error(err),
+				)
+				return nil, fmt.Errorf("applying tx %#x in block %d to worst-case state: %v", tx.Hash(), b.Height(), err)
+			}
 		}
-		if err := clock.AfterBlock(lim, vm.hooks, b.Header()); err != nil {
-			return nil, err
+		for i, op := range vm.hooks.EndOfBlockOps(b.EthBlock()) {
+			if err := state.Apply(op); err != nil {
+				log.Warn("Could not apply op during historical worst case calculation",
+					zap.Int("op_index", i),
+					zap.Stringer("op_id", op.ID),
+					zap.Error(err),
+				)
+				return nil, fmt.Errorf("applying op at end of block %d to worst-case state: %v", b.Height(), err)
+			}
+		}
+		if err := state.FinishBlock(); err != nil {
+			log.Warn("Could not finish historical worst case calculation",
+				zap.Error(err),
+			)
+			return nil, fmt.Errorf("finishing worst-case state for block %d: %v", b.Height(), err)
 		}
 	}
 
+	// TODO: This header hasn't had an opportunity to be populated by,
+	// [hook.Points] but it is passed into [hook.Points.SubSecondBlockTime]. We
+	// should construct the header inside of [hook.Points].
 	hdr := &types.Header{
 		ParentHash: parent.Hash(),
-		Number:     new(big.Int).Add(parent.Number(), common.Big1),
-		Time:       unix(blockTime),
 		Root:       lastSettled.PostExecutionStateRoot(),
+		Number:     new(big.Int).Add(parent.Number(), common.Big1),
+		// GasLimit is populated after [worstcase.State.StartBlock].
+		// GasUsed should be populated after [worstcase.State.ApplyTx] has
+		// finished being called.
+		Time: unix(blockTime),
+		// BaseFee is populated after [worstcase.State.StartBlock].
+
+		// All other fields are, optionally, populated by the call to
+		// [hook.BlockBuilder.BuildBlock] after applying every
+		// [types.Transaction].
+	}
+	if err := state.StartBlock(hdr); err != nil {
+		log.Warn("Could not start worst case block calculation",
+			zap.Error(err),
+		)
+		return nil, fmt.Errorf("starting worst-case state for new block: %v", err)
 	}
-	clock.BeforeBlock(vm.hooks, hdr)
-	hdr.BaseFee = clock.BaseFee().ToBig()
 
-	var included []*types.Transaction
+	hdr.GasLimit = state.GasLimit()
+	hdr.BaseFee = state.BaseFee().ToBig()
+
+	var (
+		candidates = pendingTxs(txpool.PendingFilter{
+			BaseFee: state.BaseFee(),
+		})
+		included []*types.Transaction
+	)
 	for _, ltx := range candidates {
+		// If we don't have enough gas remaining in the block for the minimum
+		// gas amount, we are done including transactions.
+		if remainingGas := state.GasLimit() - state.GasUsed(); remainingGas < params.TxGas {
+			break
+		}
+
 		tx, ok := ltx.Resolve()
```

Comment:
Thoughts on always resolving? I don't know much about the history of the `geth` mempool and why the lazy transaction was introduced. We could have `worstcase.State` accept a `txgossip.LazyTransaction` as it includes the sender (i.e. no more need for a signer) and `txpool.LazyTransaction` includes enough to convert it to a `hook.Op` without calling `Resolve()`.

Reply:
I'm also not particularly familiar with `LazyTransaction` handling.

While I don't think it would immediately be a problem, I don't love accepting a LazyTransaction in the worstcase logic. It seems pretty fragile to be using a signer that isn't clearly guaranteed to be the same between the execution logic and the worstcase logic.

We'll need to resolve anyways if we actually include it in the block.